### PR TITLE
Add ray multiple host support

### PR DESCRIPTION
**Pull request by FanhaiLu1**
This PR enable pytorch engine multiple host on TPU POD slices. 

### MVP Goal:

The current PR is MVP version of multiple host, two goals in MPV:

1. Load weight and sharding in multiple host 

2. Compute meaningful decode result 

####  Weight and sharding

With this PR, the weight and sharding on v5e-16 (4 host , total 16 chips) use 50 of memory compared with 8 chips. It worked as expected. 
 
```
memory using 804.8 MiB / 15.7 GiB (4.990588%) on TPU_10(process=3,(2,2,0,0))
memory using 804.8 MiB / 15.7 GiB (4.990588%) on TPU_11(process=3,(3,2,0,0))
memory using 804.8 MiB / 15.7 GiB (4.990588%) on TPU_14(process=3,(2,3,0,0))
memory using 804.8 MiB / 15.7 GiB (4.990588%) on TPU_15(process=3,(3,3,0,0))

```

The weight and sharding on v5e-8 (4 host , total 8 chips)
```
memory using 1.7 GiB / 15.7 GiB (10.765424%) on TPU_4(process=0,(0,2,0,0))
memory using 1.7 GiB / 15.7 GiB (10.765424%) on TPU_5(process=0,(1,2,0,0))
memory using 1.7 GiB / 15.7 GiB (10.765424%) on TPU_6(process=0,(0,3,0,0))
memory using 1.7 GiB / 15.7 GiB (10.765424%) on TPU_7(process=0,(1,3,0,0))
```

####  Meaningful Result

With this PR, first two line of the result on v5e-16 (4 host , total 16 chips). The result are meaningful and looking good to human. 
 
```

to find purpose and fulfillment.

I believe that everyone has a unique purpose and that it is up to each individual to discover and pursue theirs.

```

first two line of  the result on v5e-8 (4 host , total 8 chips)
```
to find purpose, happiness, and fulfillment. Here are some reasons why:

1. Purpose: Having a sense of purpose gives life meaning and direction. It helps individuals set goals and work towards achieving them, which can lead to a sense of accomplishment and fulfillment.

```

### Caveats

As this is the MVP,  it's important to note that there could be some limitations in terms of performance and accuracy right now.

### CODE REVIEW

File: jetstream_pt/ray_engine_master.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    max_decode_length: int = 4096,
+    model_name="llama",
+    quantize_weights=False,
+    quantize_kv=False,
+    max_cache_length=1024,
+) -> PyTorchEngineRayMaster:
+
+  ray.init(ignore_reinit_error=True)
+  pod_name = tpu.get_current_pod_name()
+  num_hosts = tpu.get_current_pod_worker_count()
```
Comment (allenwang28) [inline jetstream_pt/ray_engine_master.py:141]: maybe we can add an assertion check here that the pod_name is not `None` and `num_hosts >= 1` ?

`(None, 0)` would be the situation if we're not running on a TPU VM at all

File: jetstream_pt/ray_engine_master.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  print(f"pod_name:{pod_name}, number of host: {num_hosts}")
+  # pylint: disable-next=all
+  engine_worker_with_tpu_resource = PyTorchEngineRayWorker.options(
+      resources={"TPU": 4}
+  )
```
Comment (allenwang28) [inline jetstream_pt/ray_engine_master.py:146]: nit - we could remove this line altogether if we did `@ray.remote(resources={"TPU": 4})` in the `PyTorchEngineRayWorker` decorator

File: jetstream_pt/ray_engine_worker.py [hunk 3]
```diff
@@ -0,0 +1,851 @@
...
+
+    for d in jax.local_devices():
+      stats = d.memory_stats()
+      used = stats["bytes_in_use"]
+      limit = stats["bytes_limit"]
+      print(
+          f"memory using {fmt_size(used)} / {fmt_size(limit)} ({used/limit:%}) on {d}"
+      )
+
+  def init_decode_state_ray(
```
Comment (allenwang28) [inline jetstream_pt/ray_engine_worker.py:283]: nit: maybe this can be `init_decode_state` and the `init_decode_state` from above can become `_init_decode_state`?

File: jetstream_pt/ray_engine_master.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      decode_state: DecodeState,
+      slot: int,
+  ) -> DecodeState:
+    all_outputs = []
+    for worker in self.engine_workers:
+      output = worker.insert_ray.remote(
+          prefix=prefix, decode_state=decode_state, slot=slot
+      )
+      all_outputs.append(output)
+    _ = ray.get(all_outputs)
```
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine_master.py:78]: In interleave serving ray multiple host, prefill doesn't return anything.

**Comment by FanhaiLu1** [comment]
> This looks great!! Mostly nits, but IIUC this same approach would work for JAX/JetStream right?

Thanks Allen for reviewing it! Yes, the Jax/Jetstream should use same approach in general, we could extract common part of code (For example: Master code could be exactly same for both Jax and Pytorch), so both Jax and Pytorch part can share same code base.

**Comment by allenwang28** [comment]
Could we also add a basic unit test for Ray? One way is to take advantage of Ray's "fake" cluster if you look at [this](https://github.com/ray-project/ray/blob/master/python/ray/tests/test_autoscaler_fake_multinode.py#L12) file for instance.

File: jetstream_pt/ray_engine_master.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  # pylint: disable-next=all
+  def init_decode_state(
+      self,
+  ) -> DecodeState:
+    all_outputs = []
+    for worker in self.engine_workers:
+      output = worker.init_decode_state_ray.remote()
+      all_outputs.append(output)
+    _ = ray.get(all_outputs)
+    return None
```
Comment (qihqi) [inline jetstream_pt/ray_engine_master.py:45]: instead of returning None, is it meaningful to return some sort of "handle" here?

So basically the reason we don't need to return is because the worker will hold on the state in a local variable. However, we can also have the worker to store the state in a list and return it's index to master

master then keep a list of ints as the "handle". 

This way on good thing is that you dont need to worry about overwrite if prefill / insert is called multiple times . and helps with disagg setup

File: jetstream_pt/ray_engine_worker.py [hunk 6]
```diff
@@ -0,0 +1,851 @@
...
+  mask: jax.Array  # [batch_size, seqlen] -inf for invalid; 0 for valid
+
+
+# NOTE model specific
+
+
+@ray.remote
+# pylint: disable-next=all
+class PyTorchEngineRayWorker:
+  """Wraps functions to the Jet Engine API format."""
```
Comment (allenwang28) [inline jetstream_pt/ray_engine_worker.py:79]: It could be helpful to add more documentation here, for instance:

```
class PyTorchEngineRayWorker:
  """Ray actor representation for a PyTorch engine worker.

  PyTorchEngineRayWorker enables multi-host serving for models that exceed the memory
  capabilities of single-host TPU VMs with Ray. PyTorchEngineRayWorker should be used with a "leader" engine,
  e.g. `PyTorchRayEngine`.

  Note: For `PyTorchRayEngine` to return consistent results, it's important that `PyTorchEngineRayWorker` is able
  to maintain its state within process and only return results once its transferred to CPU device.

  """
```

(as I write this out, if we're renaming the above to `PyTorchRayEngine` then this should be `PyTorchRayWorker`)

File: jetstream_pt/ray_engine_worker.py (hunk 3 above)
Comment (allenwang28) [reply jetstream_pt/ray_engine_worker.py:283]: Ah okay that makes a lot of sense!

**Comment by FanhaiLu1** [comment]
> Could we also add a basic unit test for Ray? One way is to take advantage of Ray's "fake" cluster if you look at [this](https://github.com/ray-project/ray/blob/master/python/ray/tests/test_autoscaler_fake_multinode.py#L12) file for instance.

Looks great with Ray's "fake" cluster test. Can I add test with another PR (I tested current PR e2e manually)? I also plan to add unit test for RayWorkers in next PR.

File: jetstream_pt/ray_engine_master.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine_master.py]: Added.

File: jetstream_pt/ray_engine_master.py (hunk 5 above)
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine_master.py]: Good point! It would be very important for disaggregation serving.  The API might have large change with disaggregation serving, can we hold it for now? We will have a clear API when we do disaggregation serving.

File: install_everything.sh [hunk 7]
```diff
@@ -22,7 +22,8 @@ pip3 show libtpu-nightly && pip3 uninstall -y libtpu-nightly
 pip3 install pip install jax[tpu] -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
 # torch cpu
 pip3 install torch --index-url https://download.pytorch.org/whl/cpu
-pip3 install tensorflow flatbuffers absl-py flax sentencepiece seqio google-cloud-storage safetensors colorama coverage
+pip3 install tensorflow flatbuffers absl-py flax sentencepiece seqio google-cloud-storage 
+pip3 install safetensors colorama coverage ray[serve] humanize
```
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply install_everything.sh]: Good catch, updated.

File: jetstream_pt/ray_engine_master.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+  def generate(
+      self, params: Any, decode_state: DecodeState
+  ) -> tuple[None, engine_api.ResultTokens]:
+    all_outputs = []
+    for worker in self.engine_workers:
+      output = worker.generate_ray.remote(
+          params=params, decode_state=decode_state
+      )
+      all_outputs.append(output)
+    state, result_tokens = ray.get(all_outputs)[0]
```
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine_master.py]: Make sense, right now all the workers do all_gather.

File: jetstream_pt/ray_engine_master.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+from jetstream.engine import engine_api, tokenizer_pb2
+from jetstream_pt.ray_engine_worker import PyTorchEngineRayWorker
+
+Params = Any
+Prefix = Any
+DecodeState = Any
+
+
+class PyTorchEngineRayMaster(engine_api.Engine):
```
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine_master.py]: Updated.

File: jetstream_pt/ray_engine_master.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+  """Ray engine master to orchestrate requests and collect token response"""
+
+  def __init__(
+      self, engine_workers, tokenizer_path, context_length, batch_size
```
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine_master.py]: Agree, it would be more clear, updated.

File: jetstream_pt/ray_engine_worker.py (hunk 6 above)
Comment (FanhaiLu1) [reply jetstream_pt/ray_engine_worker.py]: Agree on renaming, updated.

**Comment by allenwang28** [comment]
> Looks great with Ray's "fake" cluster test. Can I add test with another PR (I tested current PR e2e manually)? I also plan to add unit test for RayWorkers in next PR.

Sounds good to me! Thanks for patiently addressing all of the comments, LGTM!

**Review by qihqi — APPROVED** [review]
